resources/sdk/purecloudjava/templates/ApiClient.mustache

```diff
@@ -975,4 +1031,72 @@ public class ApiClient implements AutoCloseable {
+        private long timeElapsedPerBucketInMillis = 0L;
+
+        private final List<Integer> statusCodes = Arrays.asList(429, 502, 503, 504);
+
+        public boolean shouldRetry(ApiClientConnectorResponse connectorResponse, long timeElapsedPerRequestInMillis) {
+            return totalRetryTimeInMillis < getMaxRetryTimeInMillis() && shouldWait(connectorResponse, timeElapsedPerRequestInMillis);
+        }
+
+        private boolean shouldWait(ApiClientConnectorResponse connectorResponse, long timeElapsedPerRequestInMillis) {
+            if (statusCodes.contains(connectorResponse.getStatusCode())) {
+
+                if (connectorResponse.getHeaders().containsKey("Retry-After")) {
+                    waitTimeInMillis = Integer.parseInt(connectorResponse.getHeaders().getOrDefault("Retry-After", String.valueOf(BASE_BACKOFF_TIME_IN_SEC))) * 1000L;
+                }
+                //If status code is 429 then wait until retry-after time and retry. OR If status code is retryable then for the first 5 times: wait until retry-after time and retry.
+                if (connectorResponse.getStatusCode() == 429 || retryCountBeforeBackOff++ < MAX_RETRIES_BEFORE_BACKOFF) {
+                    totalRetryTimeInMillis += (waitTimeInMillis + timeElapsedPerRequestInMillis);
+                    waitBeforeRetry(waitTimeInMillis);
+                    return true;
+                }
+                //If status code is 50x and if 3 Sec and 9 Sec buckets were already retried then wait for every 27 Sec and retry.
+                if (timeElapsedPerBucketInMillis > (FIVE_MINUTES_IN_MILLIS * 2)) {
+                    waitTimeInMillis = getWaitTimeExp(3);
+                    totalRetryTimeInMillis += (waitTimeInMillis + timeElapsedPerRequestInMillis);
+                    waitBeforeRetry(waitTimeInMillis);
+                    return true;
+                }
+                //If status code is 50x then wait for every 3 Sec and retry until 5 minutes then after wait for every 9 Sec and retry until next 5 minutes.
+                if (stopwatch == null) {
+                    stopwatch = Stopwatch.createStarted();
+                }
+
+                if (timeElapsedPerBucketInMillis < FIVE_MINUTES_IN_MILLIS) {
+                    waitTimeInMillis = getWaitTimeExp(1);
+
+                } else if (timeElapsedPerBucketInMillis < (FIVE_MINUTES_IN_MILLIS * 2)) {
+                    waitTimeInMillis = getWaitTimeExp(2);
+                }
+                waitBeforeRetry(waitTimeInMillis);
+                timeElapsedPerBucketInMillis = stopwatch.elapsed(TimeUnit.MILLISECONDS) + waitTimeInMillis + timeElapsedPerRequestInMillis;
+                totalRetryTimeInMillis += (waitTimeInMillis + timeElapsedPerRequestInMillis);
+
+                return true;
+            }
+            return false;
+        }
+
+        private void waitBeforeRetry(long waitTimeInMillis) {
+            try {
+                Thread.sleep(waitTimeInMillis);
```

Comment:
Could we log this at info?

Reply:
log added.

Reply:
I think he meant logging that the SDK will be sleeping for X duration.

Reply:
sorry, got it.